### ISSUE TITLE
Fake Trait Collections

### ISSUE THREAD

**Issue by ashfurrow** [ashfurrow/Forgeries#13]
Probably easiest using OCMock. See: https://github.com/artsy/eigen/pull/1044#discussion_r50219364

**Comment by orta** [comment]
Subclasses should do, lets avoid mocking. 
Or factory methods of known setups.

**Comment by ashfurrow** [comment]
The only reason I'm hesitant to subclass is because this is changing the behaviour of the subject itself. But I suppose partial mocks are also not great. Maybe I'll ask Twitter and see what other people do (if they write view tests at all 😅)

**Comment by orta** [comment]
I think of them as being more like size config options, and so providing known config options seems like a pretty legit use case

**Comment by ashfurrow** [comment]
Totally – I'm not talking about mocking the traits, but the property on the view, the subject under test. It's a read-only property, so mocking seemed like the least intrusive option.

**Comment by orta** [comment]
I understand, yeah, interesting.

**Comment by NachoSoto** [comment]
I think what you want is to implement `setOverrideTraitCollection(_ collection: UITraitCollection?, forChildViewController childViewController: UIViewController)` on some parent VC. You can fake it that way :)

**Comment by ashfurrow** [comment]
_Totally_ yeah, but this is specific to testing _views_ instead of view controllers. View's have their own `traitCollectionDidChange:` to lay themselves out, but the size classes it receives are `.Unspecified` until added as a part of a view controller hierarchy. 

I mean, we could just write our own method for VCs to call on views to lay themselves out, but I don't want to fight the frameworks.

**Comment by ashfurrow** [comment]
Excellent point: https://twitter.com/micheletitolo/status/690599942096687106

> we don’t unit test views. View models, yes. We run UI tests to ensure views display properly.

Could be that instead of testing the views (countdown view in our case) we just test the view controllers that use it instead.

**Comment by pietbrauer** [comment]
We had a similar problem when we started building the XING universal app. We needed to test cell layouts for iPhone 5, 6 Plus and iPad. We started by mocking `UIDevice` which quickly turned out to be a very bad idea. We ended up subclassing FBSnapshotTestCase and write something like this:

``` objective-c
- (void)testCell {
    id model = ...
    UITableViewCell *cell = ...
    cell.model = model;
    expect(cell).to.haveValidDeviceScreenshot(UIDeviceiPhone5);
    expect(cell).to.haveValidDeviceScreenshot(UIDeviceiPhone6);
    expect(cell).to.haveValidDeviceScreenshot(UIDeviceiPhoneiPad);
}
```

The specific tests where only run on the device they should be tested on. So we ran the test 3 times on different simulators. This ensured it grabbed the right assets (if you have fancy `~iPad` Screenshots or `@3x` screenshots) and also the right trait collections. In our research faking something ended just in more and more problems. I hope this helps?
